docs/tutorial/part_1.rst

```diff
@@ -0,0 +1,238 @@
+            depth = 1
+            read_only_fields = ["messages", "last_message"]
+            
+        def get_last_message(self, obj:Room):
+            return MessageSerializer(obj.messages.order_by('created_at').last()).data
+
+
+Creating the Consumers
+----------------------
+
+In the ``consumers.py`` file we will create only the room consumer for:
+    * Joining and leaving a room.
+    * Observing messages in that room.
+    * Observing the current users in the room.
+
+.. code-block:: python
+
+    import json
+    from django.shortcuts import get_object_or_404
+    from channels.generic.websocket import AsyncWebsocketConsumer
+    from channels.db import database_sync_to_async
+    from django.utils.timezone import now
+    from django.conf import settings
+    from typing import Generator
+    from djangochannelsrestframework.generics import GenericAsyncAPIConsumer, AsyncAPIConsumer
+    from djangochannelsrestframework.observer.generics import (ObserverModelInstanceMixin, action)
+    from djangochannelsrestframework.observer import model_observer
+
+    from .models import Room, Message, User
+    from .serializers import MessageSerializer, RoomSerializer, UserSerializer
+
+    class RoomConsumer(ObserverModelInstanceMixin, GenericAsyncAPIConsumer):
+        queryset = Room.objects.all()
+        serializer_class = RoomSerializer
+        lookup_field = "pk"
+
+        async def disconnect(self, code):
+            if hasattr(self, "room_subscribe"):
+                await self.remove_user_from_room(self.room_subscribe)
+                await self.notify_users()
+            await super().disconnect(code)
+
+        @action()
+        async def join_room(self, pk, **kwargs):
+            self.room_subscribe = pk
+            await self.add_user_to_room(pk)
+            await self.notify_users()
+
+        @action()
+        async def leave_room(self, pk, **kwargs):
+            await self.remove_user_from_room(pk)
+
+        @action()
+        async def create_message(self, message, **kwargs):
+            room:Room = await self.get_room(pk=self.room_subscribe)
+            await database_sync_to_async(Message.objects.create)(
+                room=room, 
+                user=self.scope["user"],
+                text=message
+            )
+
+        @action()
+        async def subscribe_to_messages_in_room(self, pk, **kwargs):
+            await self.message_activity.subscribe(room=pk)
+
+        @model_observer(Message)
+        async def message_activity(self, message, observer=None, **kwargs):
+            await self.send_json(message)
+
+        @message_activity.groups_for_signal
+        def message_activity(self, instance: Message, **kwargs):
+            yield 'room__%s' % instance.room_id
+            yield 'pk__%s' % instance.pk
+
+        @message_activity.groups_for_consumer
+        def message_activity(self, room=None, **kwargs):
+            if room is not None:
+                yield 'room__%s' % room
```

Comment:
I think might be better to use an `f` string rather than classic c-style interpolation.

Reply:
Ok, no problem!